### PR TITLE
Update and improve ReplicationController resource lifecycle test

### LINKED ISSUE #90881: Write CoreV1ReplicationController lifecycle test+promote - +7 coverage
# Progress `[8/8]`
* [x] Original approval issue: #88302 
* [x]  [APISnoop mock ticket](https://github.com/cncf/apisnoop/pull/289)
* [x]  test approval issue: [kubernetes/kubernetes#88588](https://github.com/kubernetes/kubernetes/issues/88588)
 * [x]  [update and improve](https://github.com/kubernetes/kubernetes/pull/ #90879)
* [x]  test pr : [#90880](https://github.com/kubernetes/kubernetes/pull/90880)
* [x]  two weeks soak start date: [testgrid-link](https://testgrid.k8s.io/google-gce#gci-gce-flaky&include-filter-by-regex=should%20test%20the%20lifecycle%20of%20a%20ReplicationController) Start 9 June 2020 [Master testgrid-link](https://testgrid.k8s.io/google-gce#gce-cos-master-flaky-repro&include-filter-by-regex=should%20test%20the%20lifecycle%20of%20a%20ReplicationController) 3 Flakes
* [x]  two weeks soak end date: [20 Oct 2020](https://testgrid.k8s.io/google-gce#gci-gce-flaky&include-filter-by-regex=should%20test%20the%20lifecycle%20of%20a%20ReplicationController) No Flakes for 2 weeks
* [X]  test promotion pr: [PR](https://github.com/kubernetes/kubernetes/pull/95713)

# Identify an untested feature Using APISnoop
According to this APIsnoop query, there are still some remaining ReplicationController endpoints which are untested.

```
SELECT
  operation_id,
  -- k8s_action,
  path,
  description
  FROM untested_stable_core_endpoints
  where path not like '%volume%'
  and operation_id ilike '%ReplicationController%'
 ORDER BY operation_id desc
 -- LIMIT 25
       ;
```

```
                     operation_id                      |                                path                                 |                          description                           
-------------------------------------------------------+---------------------------------------------------------------------+----------------------------------------------------------------
 replaceCoreV1NamespacedReplicationControllerStatus    | /api/v1/namespaces/{namespace}/replicationcontrollers/{name}/status | replace status of the specified ReplicationController
 readCoreV1NamespacedReplicationControllerStatus       | /api/v1/namespaces/{namespace}/replicationcontrollers/{name}/status | read status of the specified ReplicationController
 patchCoreV1NamespacedReplicationControllerStatus      | /api/v1/namespaces/{namespace}/replicationcontrollers/{name}/status | partially update status of the specified ReplicationController
 patchCoreV1NamespacedReplicationControllerScale       | /api/v1/namespaces/{namespace}/replicationcontrollers/{name}/scale  | partially update scale of the specified ReplicationController
 patchCoreV1NamespacedReplicationController            | /api/v1/namespaces/{namespace}/replicationcontrollers/{name}        | partially update the specified ReplicationController
 listCoreV1ReplicationControllerForAllNamespaces       | /api/v1/replicationcontrollers                                      | list or watch objects of kind ReplicationController
 deleteCoreV1CollectionNamespacedReplicationController | /api/v1/namespaces/{namespace}/replicationcontrollers               | delete collection of ReplicationController
(7 rows)
```

# Use API Reference to Lightly Document the Feature
* [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
* [client-go - ReplicationController](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/replicationcontroller.go)

# The mock test
## Test outline
1. Create a ReplicationController with a static label
2. Patch the ReplicationController with a new Label and updated data
3. Patch the ReplicationController's status
4. Patch/Scale the ReplicationController's Replica count to 2 Replicas
5. Get the ReplicationController
6. Replace the ReplicationController's Status
7. List all ReplicationControllers in all Namespaces find the ReplicationControllers(1) ensure that the ReplicationController is found and is patched
8. Delete Namespaced ReplicationControllers(1) via a Collection with a LabelSelector

## Example in Go
```go
package main

import (
  "encoding/json"
  "fmt"
  "flag"
  "os"

  v1 "k8s.io/api/core/v1"
  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
  "k8s.io/client-go/kubernetes"
  "k8s.io/apimachinery/pkg/runtime/schema"
  "k8s.io/apimachinery/pkg/types"
  autoscalingv1 "k8s.io/api/autoscaling/v1"
  "k8s.io/client-go/dynamic"
  "k8s.io/client-go/tools/clientcmd"
  //"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
)

func main() {
  // uses the current context in kubeconfig
  kubeconfig := flag.String("kubeconfig", fmt.Sprintf("%v/%v/%v", os.Getenv("HOME"), ".kube", "config"), "(optional) absolute path to the kubeconfig file")
  flag.Parse()
  config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
  if err != nil {
    fmt.Println(err)
  }
  // make our work easier to find in the audit_event queries
  config.UserAgent = "live-test-writing"
  // creates the clientset
  ClientSet, _ := kubernetes.NewForConfig(config)
  DynamicClientSet, _ := dynamic.NewForConfig(config)
  rcResource := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "replicationcontrollers"}

  testRcName := "rc-test"
  testRcNamespace := "default"
  testRcInitialReplicaCount := int32(1)
  testRcMaxReplicaCount := int32(2)

  rcTest := v1.ReplicationController{
      ObjectMeta: metav1.ObjectMeta{
          Name: testRcName,
          Labels: map[string]string{"test-rc-static": "true"},
      },
      Spec: v1.ReplicationControllerSpec{
          Replicas: &testRcInitialReplicaCount,
          Selector: map[string]string{"test-rc-static": "true"},
          Template: &v1.PodTemplateSpec{
              ObjectMeta: metav1.ObjectMeta{
                  Name: testRcName,
                  Labels: map[string]string{"test-rc-static": "true"},
              },
              Spec: v1.PodSpec{
                  Containers: []v1.Container{{
                      Name: testRcName,
                      Image: "nginx",
                  }},
              },
          },
      },
  }

  // Create a ReplicationController
  _, err = ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Create(&rcTest)
  if err != nil {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] created ReplicationController")

  // setup a watch for the RC
  rcWatch, err := ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Watch(metav1.ListOptions{LabelSelector: "test-rc-static=true"})
  if err != nil {
      fmt.Println(err)
      return
  }
  rcWatchChan := rcWatch.ResultChan()

  fmt.Println("[status] watching for all Replicas to be ready")
  for event := range rcWatchChan {
      rc, ok := event.Object.(*v1.ReplicationController)
      if ok != true {
          fmt.Println("Unable to fix type")
          return
      }
      if rc.Status.Replicas == testRcInitialReplicaCount && rc.Status.ReadyReplicas == testRcInitialReplicaCount {
          break
      }
  }

  rcLabelPatchPayload, err := json.Marshal(v1.ReplicationController{
      ObjectMeta: metav1.ObjectMeta{
          Labels: map[string]string{"test-rc": "patched"},
      },
  })
  if err != nil {
    fmt.Println(err)
    return
  }
  // Patch the ReplicationController
  _, err = ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Patch(testRcName, types.StrategicMergePatchType, []byte(rcLabelPatchPayload))
  if err != nil {
    fmt.Println(err)
    return
  }
  fmt.Println("[status] patched ReplicationController")

  rcStatusPatchPayload, err := json.Marshal(map[string]interface{}{
    "status": map[string]interface{}{
      "readyReplicas": 0,
      "availableReplicas": 0,
    },
  })
  if err != nil {
      fmt.Println(err)
      return
  }

  // Patch the ReplicationController's status
  rcStatus, err := ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Patch(testRcName, types.StrategicMergePatchType, []byte(rcStatusPatchPayload), "status")
  if err != nil {
      fmt.Println(err)
      return
  }
  if rcStatus.Status.ReadyReplicas != 0 {
      fmt.Println("[error] failed to patch ReplicationController status; ReadyReplicas != 0")
      return
  }
  fmt.Println("[status] patched status of ReplicationController")

  rcStatusUnstructured, err := DynamicClientSet.Resource(rcResource).Namespace(testRcNamespace).Get(testRcName, metav1.GetOptions{}, "status")
  if err != nil {
      fmt.Println(err)
      return
  }
  rcStatusUjson, _ := json.Marshal(rcStatusUnstructured)
	json.Unmarshal(rcStatusUjson, &rcStatus)
  if rcStatus.Status.Replicas != testRcInitialReplicaCount {
     fmt.Println("[error] Replica count does not match initial Replica count")
  }
  fmt.Println("[status] fetched ReplicationController status")

  rcScalePatchPayload, err := json.Marshal(autoscalingv1.Scale{
      Spec: autoscalingv1.ScaleSpec{
          Replicas: 2,
      },
  })
  if err != nil {
      fmt.Println(err)
      return
  }

  // Patch the ReplicationController's scale
  rcScale, err := ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Patch(testRcName, types.StrategicMergePatchType, []byte(rcScalePatchPayload), "scale")
  if err != nil {
      fmt.Println(err)
      return
  }
  rcScaleReplicasExpect := int32(2)
  if rcScale.Status.Replicas == rcScaleReplicasExpect {
      fmt.Println("[error] failed to patch ReplicationController scale; ReadyReplicas != 2")
      return
  }
  fmt.Println("[status] patched scale of ReplicationController")

  fmt.Println("[status] watching for all Replicas to be ready")
  for event := range rcWatchChan {
      rc, ok := event.Object.(*v1.ReplicationController)
      if ok != true {
          fmt.Println("Unable to fix type")
          return
      }
      if rc.Status.Replicas == testRcMaxReplicaCount && rc.Status.ReadyReplicas == testRcMaxReplicaCount {
          break
      }
  }

  // Get the ReplicationController
  rc, err := ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Get(testRcName, metav1.GetOptions{})
  if err != nil {
      fmt.Println(err)
      return
  }
  if rc.ObjectMeta.Labels["test-rc"] != "patched" {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] fetched ReplicationController")

  rcStatusUpdatePayload := rc
  rcStatusUpdatePayload.Status.AvailableReplicas = 1
  rcStatusUpdatePayload.Status.ReadyReplicas = 1

  // Replace the ReplicationController's status	
  rcStatus, err = ClientSet.CoreV1().ReplicationControllers(testRcNamespace).UpdateStatus(rcStatusUpdatePayload)
  if err != nil {
      fmt.Println(err)
      return
  }
  if rcStatus.Status.ReadyReplicas != 1 {
      fmt.Println("[error] failed to patch ReplicationController status; ReadyReplicas != 1")
      return
  }
  fmt.Println("[status] updated ReplicationController status")

  fmt.Println("[status] watching for all Replicas to be ready")
  for event := range rcWatchChan {
      rc, ok := event.Object.(*v1.ReplicationController)
      if ok != true {
          fmt.Println("Unable to fix type")
          return
      }
      if rc.Status.Replicas == testRcMaxReplicaCount && rc.Status.ReadyReplicas == testRcMaxReplicaCount {
          fmt.Println("[status] all Replicas are ready")
          break
      }
  }

  rcs, err := ClientSet.CoreV1().ReplicationControllers("").List(metav1.ListOptions{LabelSelector: "test-rc-static=true"})
  if err != nil {
      fmt.Println(err)
      return
  }
  if len(rcs.Items) == 0 {
      fmt.Println("[error] no ReplicationController were found")
  }
  foundRc := false
  for _, rcItem := range rcs.Items {
      if rcItem.ObjectMeta.Name == testRcName &&
         rcItem.ObjectMeta.Namespace == testRcNamespace &&
         rcItem.ObjectMeta.Labels["test-rc-static"] == "true" &&
         rcItem.ObjectMeta.Labels["test-rc"] == "patched" && 
         rcItem.Status.Replicas == testRcMaxReplicaCount &&
         rcItem.Status.ReadyReplicas == testRcMaxReplicaCount {
         foundRc = true
      }
  }
  if foundRc == false {
      fmt.Println("[error] unable to find ReplicationController")
      return
  }
  fmt.Println("[status] retrieved all ReplicationControllers selecting with LabelSelector")

  // Delete ReplicationController
  err = ClientSet.CoreV1().ReplicationControllers(testRcNamespace).DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: "test-rc-static=true"})
  if err != nil {
    fmt.Println(err)
    return
  }
  fmt.Println("[status] deleted ReplicationController")

  fmt.Println("[status] complete")

}
```

```go
[status] created ReplicationController
[status] watching for all Replicas to be ready
[status] patched ReplicationController
[status] patched status of ReplicationController
[status] fetched ReplicationController status
[status] patched scale of ReplicationController
[status] watching for all Replicas to be ready
[status] fetched ReplicationController
[status] updated ReplicationController status
[status] watching for all Replicas to be ready
[status] all Replicas are ready
[status] retrieved all ReplicationControllers selecting with LabelSelector
[status] deleted ReplicationController
[status] complete
```

# Verify Increase it Coverage with APISnoop
Discover useragents:

```
select distinct useragent from audit_event where bucket='apisnoop' and useragent not like 'kube%' and useragent not like 'coredns%' and useragent not like 'kindnetd%' and useragent like 'live%';
```

```
     useragent     
-------------------
 live-test-writing
(1 row)
```

See all endpoints hit from this test

```
select * from endpoints_hit_by_new_test where useragent like 'live%'; 
```

```
     useragent     |                     operation_id                      | hit_by_ete | hit_by_new_test 
-------------------+-------------------------------------------------------+------------+-----------------
 live-test-writing | createCoreV1NamespacedReplicationController           | t          |               2
 live-test-writing | deleteCoreV1CollectionNamespacedReplicationController | f          |               2
 live-test-writing | listCoreV1NamespacedReplicationController             | t          |               1
 live-test-writing | listCoreV1ReplicationControllerForAllNamespaces       | f          |               1
 live-test-writing | patchCoreV1NamespacedReplicationController            | f          |               2
 live-test-writing | patchCoreV1NamespacedReplicationControllerScale       | f          |               2
 live-test-writing | patchCoreV1NamespacedReplicationControllerStatus      | f          |               2
 live-test-writing | readCoreV1NamespacedReplicationController             | t          |               1
 live-test-writing | readCoreV1NamespacedReplicationControllerStatus       | f          |               1
 live-test-writing | replaceCoreV1NamespacedReplicationControllerStatus    | f          |               2
(10 rows)
```

Display test coverage change

```
select * from projected_change_in_coverage;
```

```
   category    | total_endpoints | old_coverage | new_coverage | change_in_number 
---------------+-----------------+--------------+--------------+------------------
 test_coverage |             438 |          190 |          197 |                7
(1 row)
```

# Final notes
If a test with these calls gets merged, ****test coverage will go up by 7 points****

This test is also created with the goal of conformance promotion.

### LINKED ISSUE #90957: Watch event test verification tooling
* [Ticket status](#sec-1)
* [Current problem](#sec-2)
* [Proposed solution](#sec-3)

This Issue replace #90574

# Ticket status
* [ ]  [K8s watch events test tooling #323](https://github.com/cncf/apisnoop/pull/323)
* [ ]  [Watch event test verification tooling #90574](https://github.com/kubernetes/kubernetes/issues/90574)
* [X] [PR #91416 ](https://github.com/kubernetes/kubernetes/pull/91416)
# Current problem
Watch events may be missed and the test may still pass. We need to ensure that all expected watch events are seen.

Watch events are not collected and checked at the end of the test.

# Proposed solution
In order to solve the problem, we must collect all watch events as they come in and perform a check at the end of the test to ensure that the order is correct and they contain no errors.

```go
package main

import (
  "fmt"
)

func main() {
  expectedWatchEventsAll := [][]string{
    {
      "ADDED",
      "ADDED",
      "DELETED",
    },
    {
      "ADDED",
      "MODIFIED",
      "MODIFIED",
      "DELETED",
    },
    {
      "ADDED",
      "MODIFIED",
      "DELETED",
    },
  }
  for expectedExampleInt, expectedWatchEvents := range expectedWatchEventsAll {
	  fmt.Println("Running example", expectedExampleInt)
    allowedAttempts := 3
    failure := VerifyWatchEventOrder(allowedAttempts, expectedWatchEvents, func() []string {
      var myWatchEvent string
      var watchEvent []string
      fmt.Println("An event takes place (1/3)")
      myWatchEvent = "ADDED"
      watchEvent = append(watchEvent, myWatchEvent)

      fmt.Println("An event takes place (2/3)")
      myWatchEvent = "MODIFIED"
      watchEvent = append(watchEvent, myWatchEvent)

      fmt.Println("An event takes place (3/3)")
      myWatchEvent = "DELETED"
      watchEvent = append(watchEvent, myWatchEvent)

      return watchEvent
    })
    if failure != "" {
      fmt.Println(failure, "watch events occured in the wrong or incorrect order")
    }
    fmt.Println("Check complete")
  }
}

func VerifyWatchEventOrder(retries int, expectedWatchEvents []string, scenario func() []string) (failure string) {
  var attemptContainsFailure bool
attempts:
  for try := 1; try <= retries; try++ {
    if try == retries {
      return failure
    }
    actualWatchEvents := scenario()
    if len(expectedWatchEvents) != len(actualWatchEvents) {
      failure = "expected watch events count does not match the count of actual watch events"
      attemptContainsFailure = true
      continue
    }
    for watchEventInt, watchEvent := range actualWatchEvents {
      if expectedWatchEvents[watchEventInt] != watchEvent {
        failure = fmt.Sprintf("(index %v) %v not found, found %v instead. %v/%v attempts", watchEventInt, expectedWatchEvents[watchEventInt], watchEvent, try, retries)
        attemptContainsFailure = true
      }
    }
    if attemptContainsFailure == false {
      failure = ""
      break attempts
    }
  }
  return failure
}
```

```
Running example 0
An event takes place (1/3)
An event takes place (2/3)
An event takes place (3/3)
An event takes place (1/3)
An event takes place (2/3)
An event takes place (3/3)
(index 1) ADDED not found, found MODIFIED instead. 2/3 attempts watch events occured in the wrong or incorrect order
Check complete
Running example 1
An event takes place (1/3)
An event takes place (2/3)
An event takes place (3/3)
An event takes place (1/3)
An event takes place (2/3)
An event takes place (3/3)
expected watch events count does not match the count of actual watch events watch events occured in the wrong or incorrect order
Check complete
Running example 2
An event takes place (1/3)
An event takes place (2/3)
An event takes place (3/3)
Check complete
```

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Updates the test to rely on responses from `Patch` or `Update` instead of later `Get` or `List`

**Which issue(s) this PR fixes**:
Fixes: #90881
Resolves #90957 

 [#88588 (comment)](https://github.com/kubernetes/kubernetes/pull/88588#issuecomment-607366847)

**Special notes for your reviewer**:
Attempts to fix flakiness.

**Does this PR introduce a user-facing change?**:

```
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```
NONE
```

/sig testing

### CODE REVIEW

**Comment by Riaankl** [comment]
This PR replaces #89746

File: test/e2e/apps/rc.go [hunk 1]
```diff
@@ -139,14 +140,27 @@ var _ = SIGDescribe("ReplicationController", func() {
...
 
+		ginkgo.By("waiting for RC to be added")
+		eventFound := false
+		for watchEvent := range rcWatchChan {
+			if watchEvent.Type == watch.Added {
+				eventFound = true
+				break
+			}
+		}
+		framework.ExpectEqual(eventFound, true, "failed to find RC %v event", watch.Added)
```
Comment (spiffxp) [inline test/e2e/apps/rc.go:151]: nit: for the number of times this is copy-pasted around, I would pull this into a function
```
expectWatchEventType(rcWatch, watch.Added)
```

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apimachinery/watch.go#L397-L431 has something to crib from

File: test/e2e/apps/rc.go [hunk 2]
```diff
@@ -172,6 +196,27 @@ var _ = SIGDescribe("ReplicationController", func() {
...
+		for watchEvent := range rcWatchChan {
+			rc, ok := watchEvent.Object.(*v1.ReplicationController)
+			framework.ExpectEqual(ok, true, "Unable to convert type of ReplicationController watch watchEvent")
+
+			if rc.Status.Replicas == testRcInitialReplicaCount {
+				eventFound = true
+				break
+			}
+		}
+		framework.ExpectEqual(eventFound, true, "RC does not have initial replica count")
```
Comment (spiffxp) [inline test/e2e/apps/rc.go:219]: Why do we need this check / what is this trying to test?  

If this is trying to wait until the RC status has been updated to undo what we just patched in, it should be looking at `rc.Status.ReadyReplicas`

File: test/e2e/apps/rc.go [hunk 3]
```diff
@@ -193,6 +238,16 @@ var _ = SIGDescribe("ReplicationController", func() {
...
+		ginkgo.By("waiting for RC to be modified")
+
+		eventFound = false
+		for watchEvent := range rcWatchChan {
+			if watchEvent.Type == watch.Modified {
+				eventFound = true
+				break
+			}
+		}
+		framework.ExpectEqual(eventFound, true, "failed to find RC %v event", watch.Modified)
```
Comment (spiffxp) [inline test/e2e/apps/rc.go:250]: Do we need this?  It seems like the below watch event loop is going to wait until the RC status converges to the new number of replicas, I'm not sure why we need to wait before we wait

File: test/e2e/apps/rc.go [hunk 4]
```diff
@@ -245,9 +301,7 @@ var _ = SIGDescribe("ReplicationController", func() {
 			if rcItem.ObjectMeta.Name == testRcName &&
 				rcItem.ObjectMeta.Namespace == testRcNamespace &&
 				rcItem.ObjectMeta.Labels["test-rc-static"] == "true" &&
-				rcItem.ObjectMeta.Labels["test-rc"] == "patched" &&
-				rcItem.Status.Replicas == testRcMaxReplicaCount &&
```
Comment (spiffxp) [inline test/e2e/apps/rc.go:249]: should we be checking `Spec.Replicas`?

**Review by spiffxp — COMMENTED** [review]
/ok-to-test
/release-note-none

It's been a while since I've looked at this, and I've got to admit the test is now so long I'm having a tough time keeping track of what is being done when and why.  I'm not sure why all of the watchEvent checks have been added.  Do we know they help?

**Comment by BobyMCbobs** [comment]
This PR has been updated to use and now depends on the merging of the new [watch tooling](https://github.com/kubernetes/kubernetes/pull/91416)

File: test/e2e/apps/rc.go (hunk 1 above)
Comment (BobyMCbobs) [reply test/e2e/apps/rc.go]: Using `framework.WatchUntilWithoutRetry` (introduced in #91416) for these checks as of 313f0341ca3cc3de91c664e6f5d001dab1a6852e

**Comment by spiffxp** [comment]
/release-note-none

/approve
/lgtm
to get us some signal, I'm still not convinced this is the right way to go, will summarize in another comment

**Pull request merged**